### PR TITLE
osd/scrub: mark PG as being scrubbed, from scrub initiation to Inactive state

### PR DESCRIPTION
**Pull request by ronen-fr**
The scrubber's state-machine changes states only following a message dispatched
via the OSD queue. That creates some vulnerability periods, from when the
decision to change the state is made, till when the message carrying the event
is dequeued and processed by the state-machine.
    
One of the problems thus created is a second scrub being started on a PG, before
the previous scrub is fully terminated and cleaned up.
    
Here we add a 'being-scrubbed' flag, that is asserted when the first scrub
initiation message is queued and is only cleared when the state machine reaches
Inactive state after the scrub is done.
    
To note:
- scrub_finish() is now part of the FSM transition from WaitDigest to Inactive,
  closing this specific vulnerability period;
    
Signed-off-by: Ronen Friedman <rfriedma@redhat.com>

### CODE REVIEW

**Comment by ronen-fr** [comment]
latest round of tests: 125 tests (mostly thrashers), 3 failures:
- rados-api "regular" failure
- ceph-objetstore-tool crashed
- 6367305: does not seem related

File: src/osd/PG.cc [hunk 1]
```diff
@@ -1571,8 +1563,14 @@ void PG::on_role_change() {
...
 
 void PG::on_new_interval() {
-  dout(20) << __func__ << " scrub_queued was " << scrub_queued << " flags: " << m_planned_scrub << dendl;
-  scrub_queued = false;
+  // log some scrub data before we react to the interval
+  if (m_scrubber) {
+    dout(20) << __func__ << (m_scrubber->is_being_scrubbed() ? " scrubbing " : " ")
+             << "flags: " << m_planned_scrub << dendl;
+  } else {
+    dout(20) << __func__ << dendl;
```
Comment (neha-ojha) [inline src/osd/PG.cc:1571]: do we need this, can this not be inferred?

File: src/osd/PG.cc [hunk 2]
```diff
@@ -433,19 +432,16 @@ void PG::queue_scrub_after_repair()
...
 
-  if (is_scrubbing()) {
-    dout(10) << __func__ << ": scrubbing already" << dendl;
-    return;
-  }
-  if (scrub_queued) {
-    dout(10) << __func__ << ": already queued" << dendl;
+  if (m_scrubber->is_being_scrubbed()) {
+    dout(5) << __func__ << ": scrubbing already "
+            << " is_scrubbing: " << is_scrubbing()<< dendl;
```
Comment (neha-ojha) [inline src/osd/PG.cc:437]: why do we need this at level 5?

File: src/osd/PG.cc [hunk 3]
```diff
@@ -760,7 +756,7 @@ void PG::set_probe_targets(const set<pg_shard_t> &probe_set)
 
 void PG::send_cluster_message(
   int target, MessageRef m,
-  epoch_t epoch, bool share_map_update=false)
+  epoch_t epoch, bool share_map_update)
```
Comment (neha-ojha) [inline src/osd/PG.cc:759]: why?

File: src/osd/PG.cc [hunk 4]
```diff
@@ -2539,7 +2539,7 @@ std::pair<ghobject_t, bool> PG::do_delete_work(
       epoch_t e = get_osdmap()->get_epoch();
       PGRef pgref(this);
       auto delete_requeue_callback = new LambdaContext([this, pgref, e](int r) {
-        dout(20) << __func__ << " wake up at "
+        dout(20) << "do_delete_work() [cb] wake up at "
```
Comment (neha-ojha) [inline src/osd/PG.cc:2542]: This change isn't relevant and can go in a separate commit.

**Review by neha-ojha — COMMENTED** [review]
some early comments, still reviewing

File: src/osd/PG.cc (hunk 4 above)
Comment (ronen-fr) [reply src/osd/PG.cc:2542]: Fixed already

File: src/osd/PG.cc (hunk 2 above)
Comment (ronen-fr) [reply src/osd/PG.cc:437]: debugging. Will revert.

File: src/osd/PG.cc (hunk 3 above)
Comment (ronen-fr) [reply src/osd/PG.cc:759]: It's a virtual function, and as such - better not have optional parameters marked by the derivatives.
(I would remove such optional flags whenever possible, any way. They are a relic of the times when we
didn't have overloads in C++, and create more problems than they help)

File: src/osd/PG.cc (hunk 1 above)
Comment (ronen-fr) [reply src/osd/PG.cc:1571]: I do wish to know when a new interval starts but the PG has no Scrubber object. Added as I needed that
during debugging.

**Comment by ronen-fr** [comment]
Must fix non-intentional submodule changes

**Comment by ronen-fr** [comment]
DNM for now. Latest rebase not really working yet.

**Comment by ronen-fr** [comment]
The 'not signed' failure seems to be caused by "no space left on device".

**Comment by ronen-fr** [comment]
jenkins test signed please

**Comment by ronen-fr** [comment]
jenkings retest this please

File: src/osd/PG.cc [hunk 5]
```diff
@@ -1331,15 +1327,12 @@ Scrub::schedule_result_t PG::sched_scrub()
 	  << (is_clean() ? " <clean>" : " <not-clean>") << dendl;
   ceph_assert(ceph_mutex_is_locked(_lock));
 
-  if (!is_primary() || !is_active() || !is_clean()) {
-    return Scrub::schedule_result_t::bad_pg_state;
+  if (!m_scrubber || m_scrubber->is_being_scrubbed()) {
```
Comment (athanatos) [inline src/osd/PG.cc:1330]: Can m_scrubber actually be null here?

Comment (athanatos) [reply src/osd/PG.cc:1330]: PG::m_scrubber should probably have a comment explaining that it's only null during construction and destruction.

File: src/osd/PG.cc (hunk 3 above)
Comment (athanatos) [reply src/osd/PG.cc:759]: Move this to its own commit.

File: src/osd/PG.cc [hunk 6]
```diff
@@ -633,8 +629,8 @@ void PG::release_backoffs(const hobject_t& begin, const hobject_t& end)
       auto q = p->second.begin();
       while (q != p->second.end()) {
 	dout(20) << __func__ << " checking  " << *q << dendl;
-	int r = cmp((*q)->begin, begin);
-	if (r == 0 || (r > 0 && (*q)->end < end)) {
+	int rr = cmp((*q)->begin, begin);
+	if (rr == 0 || (rr > 0 && (*q)->end < end)) {
```
Comment (athanatos) [inline src/osd/PG.cc:633]: Move this to its own commit.

File: src/osd/PG.cc [hunk 7]
```diff
@@ -2092,6 +2092,9 @@ void PG::forward_scrub_event(ScrubAPI fn, epoch_t epoch_queued, std::string_view
   if (is_active() && m_scrubber) {
     ((*m_scrubber).*fn)(epoch_queued);
   } else {
+    if (m_scrubber) {
```
Comment (athanatos) [inline src/osd/PG.cc:2095]: This should probably be an assert, shouldn't be possible to invoke forward_scrub_event on a destructed PG.

File: src/osd/PG.cc [hunk 8]
```diff
@@ -1587,8 +1581,14 @@ void PG::on_role_change() {
 }
 
 void PG::on_new_interval() {
-  dout(20) << __func__ << " scrub_queued was " << scrub_queued << " flags: " << m_planned_scrub << dendl;
-  scrub_queued = false;
+  // log some scrub data before we react to the interval
+  if (m_scrubber) {
```
Comment (athanatos) [inline src/osd/PG.cc:1585]: This should probably be an assert, shouldn't be possible to invoke on_new_interval() on a destructed PG.

File: src/osd/scrubber/pg_scrubber.h [hunk 9]
```diff
@@ -415,6 +418,13 @@ class PgScrubber : public ScrubPgIF, public ScrubMachineListener {
 
   [[nodiscard]] bool was_epoch_changed() const final;
 
+  void set_being_scrubbed() final;
+  void clear_being_scrubbed() final;
+
+  void set_finishing_flag() final { m_finish_sequence_started = true; }
+  void clear_finishing_flag() final { m_finish_sequence_started = false; }
+  bool is_finishing_flag_set() const final { return m_finish_sequence_started; }
```
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.h:426]: I don't really understand what's going on here.

File: src/osd/scrubber/pg_scrubber.h [hunk 10]
```diff
@@ -660,6 +685,16 @@ class PgScrubber : public ScrubPgIF, public ScrubMachineListener {
...
 
+  /**
+   *  preventing WaitDigestUpdate state from reacting to DigestUpdate events
+   *  once scrub_finish() is called.
+   *  Note - making this a separate FSM state would have opened the same race window
+   *  we are trying to minimize. There is no safe way to have the non-reentrant Boost
+   *  state-chart transition into a new state without going through the queue (which
+   *  also means releasing the PG lock).
+   */
+  bool m_finish_sequence_started{false};
```
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.h:696]: I don't think I understand how we can get a stray DigestUpdate.

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/osd/PG.cc (hunk 5 above)
Comment (ronen-fr) [reply src/osd/PG.cc]: fixed

File: src/osd/PG.cc (hunk 6 above)
Comment (ronen-fr) [reply src/osd/PG.cc:633]: done

File: src/osd/PG.cc (hunk 8 above)
Comment (ronen-fr) [reply src/osd/PG.cc]: fixed

File: src/osd/scrubber/pg_scrubber.h (hunk 9 above)
Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.h]: removed all that 'finishing flag' stuff.

File: src/osd/scrubber/pg_scrubber.h (hunk 10 above)
Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.h]: removed all that 'finishing flag' stuff. I made a minor change to the logic that sens the event, and hope it really won't happen.

File: src/osd/PG.cc (hunk 7 above)
Comment (ronen-fr) [reply src/osd/PG.cc]: done

**Comment by ronen-fr** [comment]
jenkins retest this please

File: src/osd/PG.cc [hunk 11]
```diff
@@ -1337,16 +1332,14 @@ Scrub::schedule_result_t PG::sched_scrub()
 	  << (is_active() ? ") <active>" : ") <not-active>")
 	  << (is_clean() ? " <clean>" : " <not-clean>") << dendl;
   ceph_assert(ceph_mutex_is_locked(_lock));
+  ceph_assert(m_scrubber);
 
-  if (!is_primary() || !is_active() || !is_clean()) {
-    return Scrub::schedule_result_t::bad_pg_state;
+  if (is_scrub_queued_or_active()) {
+    return Scrub::schedule_result_t::already_started;;
```
Comment (athanatos) [inline src/osd/PG.cc:1338]: Extra ;

File: src/osd/PG.cc [hunk 12]
```diff
@@ -2101,10 +2100,12 @@ void PG::repair_object(
 void PG::forward_scrub_event(ScrubAPI fn, epoch_t epoch_queued, std::string_view desc)
 {
   dout(20) << __func__ << ": " << desc << " queued at: " << epoch_queued << dendl;
-  if (is_active() && m_scrubber) {
+  ceph_assert(m_scrubber);
+  if (is_active()) {
     ((*m_scrubber).*fn)(epoch_queued);
   } else {
     // pg might be in the process of being deleted
+    m_scrubber->clear_queued_or_active();
```
Comment (athanatos) [inline src/osd/PG.cc:2108]: Why does this cause a state change in the scrubber?

Comment (ronen-fr) [reply src/osd/PG.cc]: fixed

**Comment by ronen-fr** [comment]
Following a discussion with Sam Just: a follow-up PR will reorder and simplify the
'is this scrub message still relevant' checks.

**Review by athanatos — APPROVED** [review]
LGTM.  Probably should retest prior to merge?

**Comment by ronen-fr** [comment]
@athanatos - thanks.

> Probably should retest prior to merge?

For sure

**Comment by ronen-fr** [comment]
https://pulpito.ceph.com/?branch=wip-ronenf-unique-scrub
107 tests for the latest version: 1 known failure

About 200 addition tests of the last 3 versions, which mostly differ in immaterial details.
Merging based on these (and previous) results.
